plugins/BEdita/Core/src/Utility/JsonSchema.php

```diff
@@ -0,0 +1,135 @@
+ */
+class JsonSchema
+{
+    /**
+     * Valid resource names
+     *
+     * @var array
+     */
+    const VALID_RESOURCES = [
+        'applications',
+        'roles',
+        'streams',
+    ];
+
+    /**
+     * Generate JSON Schema draft-6 (https://tools.ietf.org/html/draft-wright-json-schema-01)
+     *
+     * @param string $typeName Type name of a resource or object
+     * @param string $url Url of this schema
+     * @return mixed
+     */
+    public static function generate($typeName, $url)
+    {
+        $schema = static::typeSchema($typeName);
+        if (!is_array($schema)) {
+            return $schema;
+        }
+
+        $baseSchema = [
+            'definitions' => new \stdClass(),
+            '$id' => $url,
+            '$schema' => 'http://json-schema.org/draft-06/schema#',
+            'type' => 'object',
+        ];
+
+        return array_merge($baseSchema, $schema);
+    }
+
+    /**
+     * Schema of a resource or object type
+     *
+     * @param string $typeName Resource or object type name
+     * @return mixed
+     * @throws \Cake\Network\Exception\NotFoundException if no type is found
+     */
+    public static function typeSchema($typeName)
+    {
+        if (in_array($typeName, static::VALID_RESOURCES)) {
+            return static::resourceSchema($typeName);
+        }
+
+        /* @var \BEdita\Core\Model\Table\ObjectTypesTable $ObjectTypes */
+        $ObjectTypes = TableRegistry::get('ObjectTypes');
+        try {
+            $objectType = $ObjectTypes->get($typeName);
+
+            return static::objectSchema($objectType);
+        } catch (RecordNotFoundException $e) {
+            throw new NotFoundException(__d('bedita', 'Type "{0}" not found', $typeName));
+        }
+    }
+
+    /**
+     * Build resource properties directly from a db table
+     *
+     * @param string $resource Resource type name
+     * @return array JSON Schema array with `properties` and `required`
+     */
+    public static function resourceSchema($resource)
+    {
+        $table = TableRegistry::get((string)Inflector::camelize($resource));
+        $entity = $table->newEntity();
+        $schema = $table->getSchema();
+        $hiddenProperties = $entity->hiddenProperties();
+
+        $properties = [];
+        $required = [];
+        foreach ($schema->columns() as $name) {
+            if (in_array($name, (array)$table->getPrimaryKey()) || in_array($name, $hiddenProperties)) {
+                continue;
+            }
+
+            $accessMode = null;
+            if (!$entity->isAccessible($name)) {
+                $accessMode = 'readOnly';
+            } elseif (in_array($name, $hiddenProperties)) {
+                $accessMode = 'writeOnly';
+            }
+            $property = new StaticProperty(compact('name', 'table'));
+            $properties[$name] = $property->getSchema($accessMode);
+            if ($property->required && $accessMode !== 'readOnly') {
```

Comment:
`writeOnly` hidden properties could be required ? (same note in `Model/Entity/ObjectType.php`)

Reply:
Same question as above also for me :recycle:

Reply:
Right now, yes. But I see that it might be confusing at times… 🤔